### PR TITLE
[REEF-802] Fix medium-priority warnings reported by Findbugs, part 2

### PR DESCRIPTION
**Pull request by tcNickolas**
NOTE: Behavior changes are small and similar to https://github.com/apache/reef/pull/756, but they are there. Please review carefully.
This does close REEF-802

This change fixes or configures to be ignored meduim-priority bugs
in all Java code and enforces medium-priority bugs to break the build.

JIRA:
  [REEF-802](https://issues.apache.org/jira/browse/REEF-802)

Pull request:
  This closes #

### CODE REVIEW

File: lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java [hunk 1]
```diff
@@ -62,6 +62,24 @@ public String toString() {
...
+  public boolean equals(final Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    NumberedSplit<?> that = (NumberedSplit<?>) o;
+    return index == that.index;
```
Comment (dongjoon-hyun) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java]: Hi, @tcNickolas . Is it enough to compare only `index`? `NumberedSplit` has two more members.
```
new NumberedSplit<>(split, splitNum, partitions[splitNum]);
```

File: lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java [hunk 2]
```diff
@@ -62,6 +62,24 @@ public String toString() {
...
+  }
+
+  @Override
+  public int hashCode() {
+    return index;
```
Comment (dongjoon-hyun) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java]: The other members, too?

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/examples/accumulate/CombinerStage.java [hunk 3]
```diff
@@ -119,6 +119,24 @@ public Pair(final K k, final V v) {
...
+    public boolean equals(final Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      Pair<K, V> pair = (Pair<K, V>) o;
+      return k.compareTo(pair.getKey()) == 0;
```
Comment (dongjoon-hyun) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/examples/accumulate/CombinerStage.java]: `pair.getValue()`?

File: lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java (hunk 1 above)
Comment (tcNickolas) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java]: Results of `equals()` must match results of `compareTo()`. In this case `compareTo()` is implemented based on index comparison only, so I can't use any other fields in `equals()`, or there will be mismatches.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java (hunk 2 above)
Comment (tcNickolas) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java]: And `hashCode()` must match the results of other two, so members which are considered equal must have same hash code.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java (hunk 1 above)
Comment (dongjoon-hyun) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java]: Then, why don't you change them both? Is the current logic correct?

Comment (tcNickolas) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java]: We never compare splits from different sources, and splits from one source can be compared based on index, so yes, it looks correct to me.

Comment (dongjoon-hyun) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/NumberedSplit.java]: Then, okay. This PR passed on Mac/Windows and Yarn(on Docker). As you said, current implementation (including yours) looks enough and correct for now. If there is any problem in the future, we can fix at that time. I'll merge this. Thank you, @tcNickolas .

**Pull request closed without merging**